File: data-structures/linked-list/linked-list-class/java/LinkedList.java
/**
 * LinkedList is a class that represents a singly linked list data structure.
 * It allows insertion of nodes at specific positions, searching for nodes, and
 * deletion of nodes.
 */
public class LinkedList {
    private Node head;
    private int size;

    /**
     * Constructs an empty LinkedList with a null head and size 0.
     */
    public LinkedList() {
        this.head = null;
        this.size = 0;
    }

    /**
     * Returns the size of the LinkedList.
     *
     * @return The size of the LinkedList.
     */
    public int getSize() {
        return this.size;
    }

    /**
     * Inserts a new node with the specified value at the given index.
     * Throws an IndexOutOfBoundsException if the index is out of range.
     *
     * @param value The value to be inserted.
     * @param index The index at which the value should be inserted.
     * @throws IndexOutOfBoundsException if the index is out of range.
     */
    public void insertData(String value, int index) {
        if (index < 0 || index > size) {
            throw new IndexOutOfBoundsException("Index out of range");
        }

        // Insert at the beginning
        if (index == 0) {
            Node node = new Node(value, this.head);
            this.head = node;
            this.size += 1;
            return;
        }

        // Insert at a specific index
        Node node = new Node(value, null);
        Node iterNode = this.head;
        int i = 0;
        while (i < index - 1) {
            iterNode = iterNode.getNext();
            i++;
        }
        // Insert at the end
        if (index == this.getSize()) {
            iterNode.setNext(node);
            this.size += 1;
            return;
        }
        // Insert in the middle
        node.setNext(iterNode.getNext());
        iterNode.setNext(node);
        this.size += 1;
    }

    /**
     * Searches for a node with the specified value in the LinkedList.
     *
     * @param value The value to search for.
     * @return true if the value is found, false otherwise.
     */
    public boolean find(String value) {
        int i = 0;
        Node iterNode = this.head;
        while (i < this.getSize()) {
            if (iterNode.getValue().equals(value))
                return true;
            i++;
            iterNode = iterNode.getNext();
        }
        return false;
    }

    /**
     * Deletes the node at the given index.
     * Throws an IndexOutOfBoundsException if the index is out of range.
     *
     * @param index The index of the node to delete.
     * @throws IndexOutOfBoundsException if the index is out of range.
     */
    public void deleteData(int index) {
        if (index < 0 || index > size - 1) {
            throw new IndexOutOfBoundsException("Index out of range");
        }
        int i = 0;
        Node iterNode = this.head;
        while (i < index - 1) {
            iterNode = iterNode.getNext();
            i++;
        }
        if (index == this.getSize()) {
            iterNode.setNext(null);
            this.size -= 1;
            return;
        }
        iterNode.setNext(iterNode.getNext().getNext());
        this.size -= 1;
    }
}

File: data-structures/linked-list/linked-list-class/python/linked_list.py
from node import Node


class LinkedList:
    def __init__(self) -> None:
        """
        Initializes an empty linked list.
        """
        self._head = None
        self._size = 0

    def get_size(self) -> int:
        """
        Returns the size of the linked list.

        Returns:
            The number of nodes in the linked list.
        """
        return self._size

    def insert_node(self, value: str, index: int) -> None:
        """
        Inserts a new node with the given value at the specified index.

        Args:
            value (str): The value to be inserted.
            index (int): The index at which the value should be inserted.

        Raises:
            ValueError: If the index is out of range or the list is empty (index >= 0).

        """
        if index < 0 or index > self._size:
            raise ValueError('Index out of range')

        if self._head == None and index > 0:
            raise ValueError('List is empty, index out of range')

        node = Node(value)
        if index == 0:
            node.set_next(self._head)
            self._head = node
            self._size += 1
        else:
            iterNode = self._head
            i = 0
            while i < index - 1:
                iterNode = iterNode.get_next()
                i += 1

            if index == self._size:
                iterNode.set_next(node)
                self._size += 1
                return

            node.set_next(iterNode.get_next())
            iterNode.set_next(node)
            self._size += 1

    def find(self, value: str) -> bool:
        """
        Searches for a node with the given value in the linked list.

        Args:
            value (str): The value to search for.

        Returns:
            bool: True if the value is found, False otherwise.

        Raises:
            ValueError: If the list is empty.
        """
        if self._head == None:
            raise ValueError("List is empty")

        node = self._head
        while node != None:
            if node.get_value() == value:
                return True
            node = node.get_next()
        return False

    def deleteNode(self, index: int) -> None:
        """
        Deletes the node at the specified index.

        Args:
            index (int): The index of the node to be deleted.

        Raises:
            ValueError: If the index is out of range or the list is empty (index >= 0).
        """
        if index < 0 or index > self._size:
            raise ValueError('Index out of range')

        if self._head == None:
            raise ValueError('List is empty, index out of range')

        if index == 0:
            self._head = self._head.get_next()
            self._size -= 1
            return

        node = self._head
        i = 0
        while i < index - 1:
            node = node.get_next()
            i += 1

        if index == self._size - 1:
            node.set_next(None)
            self._size -= 1
            return

        node.set_next(node.get_next().get_next())
        self._size -= 1
